NitroxServer/GameLogic/JoiningManager.cs

```diff
@@ -0,0 +1,236 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NitroxModel.DataStructures;
+using NitroxModel.DataStructures.GameLogic;
+using NitroxModel.DataStructures.GameLogic.Entities;
+using NitroxModel.DataStructures.Unity;
+using NitroxModel.DataStructures.Util;
+using NitroxModel.MultiplayerSession;
+using NitroxModel.Packets;
+using NitroxModel.Serialization;
+using NitroxServer.Communication;
+using NitroxServer.GameLogic.Bases;
+using NitroxServer.Serialization.World;
+
+namespace NitroxServer.GameLogic;
+
+public sealed class JoiningManager
+{
+    private readonly PlayerManager playerManager;
+    private readonly SubnauticaServerConfig serverConfig;
+    private readonly World world;
+    private readonly SessionSettings sessionSettings;
+
+    private ThreadSafeQueue<(INitroxConnection, string)> joinQueue { get; } = new();
+    private readonly Lock queueLocker = new(); // Necessary to avoid race conditions between JoinQueueLoop and AddToJoinQueue
+    private bool queueActive;
+    public Action SyncFinishedCallback { get; private set; }
+
+    public JoiningManager(PlayerManager playerManager, SubnauticaServerConfig serverConfig, World world, SessionSettings sessionSettings)
+    {
+        this.playerManager = playerManager;
+        this.serverConfig = serverConfig;
+        this.world = world;
+        this.sessionSettings = sessionSettings;
+    }
+
+    private async Task JoinQueueLoop()
+    {
+        // It may be possible to use the task's status itself for this,
+        // but the ContinueWithHandleError callback might cause issues
+        queueActive = true;
+
+        while (true)
+        {
+            lock (queueLocker)
+            {
+                if (joinQueue.Count == 0)
+                {
+                    queueActive = false;
+                    return;
+                }
+            }
+
+            try
+            {
+                (INitroxConnection connection, string reservationKey) = joinQueue.Dequeue();
+                string name = playerManager.GetPlayerContext(reservationKey).PlayerName;
+
+                // Do this after dequeueing because everyone's position shifts forward
+                (INitroxConnection, string)[] array = [.. joinQueue];
+                for (int i = 0; i < array.Length; i++)
+                {
+                    (INitroxConnection c, _) = array[i];
+                    c.SendPacket(new JoinQueueInfo(i + 1, serverConfig.InitialSyncTimeout));
+                }
+
+                Log.Info($"Starting sync for player {name}");
+                SendInitialSync(connection, reservationKey);
+
+                CancellationTokenSource source = new(serverConfig.InitialSyncTimeout);
```

Comment:
Should dispose of CTS.

```suggestion
                using CancellationTokenSource source = new(serverConfig.InitialSyncTimeout);
```